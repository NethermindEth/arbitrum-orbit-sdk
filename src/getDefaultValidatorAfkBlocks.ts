import { Client, Transport, Chain } from 'viem';

import { ParentChainId, validateParentChain } from './types/ParentChain';
import { parentChainIsMainnet } from './parentChainIsMainnet';
import { getParentChainBlockTime } from './getParentChainBlockTime';

export function getDefaultValidatorAfkBlocks<TChain extends Chain | undefined>(
  parentChainIdOrClient: ParentChainId | Client<Transport, TChain>,
): bigint {
  const { chainId: parentChainId, isCustom: parentChainIsCustom } =
    validateParentChain(parentChainIdOrClient);

  if (parentChainIsCustom) {
    throw new Error(
      `[getDefaultValidatorAfkBlocks] can't provide defaults for custom parent chain with id ${parentChainId}`,
    );
  }

  const isMainnet = parentChainIsMainnet(parentChainId);
  const blocksPerMinute = 60 / getParentChainBlockTime(parentChainId);

  return isMainnet
    ? // 28 days
      BigInt(28 * 24 * 60 * blocksPerMinute)
    : // 7 days
      BigInt(7 * 24 * 60 * blocksPerMinute);
}
