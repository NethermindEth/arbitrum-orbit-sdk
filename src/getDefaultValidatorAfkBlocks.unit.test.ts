import { it, expect } from 'vitest';

import { chains } from './chains';
import { ParentChainId } from './types/ParentChain';
import { getDefaultValidatorAfkBlocks } from './getDefaultValidatorAfkBlocks';

it('returns default value for validatorAfkBlocks based on parent chain', () => {
  expect(
    chains
      .filter((chain) => chain.id !== 333333)
      .reduce((acc, value) => {
        return {
          ...acc,
          // it's ok to cast as we've filtered out 333333 above
          [value.id]: getDefaultValidatorAfkBlocks(value.id as ParentChainId),
        };
      }, {}),
  ).toMatchSnapshot();
});
