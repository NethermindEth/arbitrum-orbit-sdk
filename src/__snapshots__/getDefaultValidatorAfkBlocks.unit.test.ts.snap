// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns default value for validatorAfkBlocks based on parent chain 1`] = `
{
  "1": 201600n,
  "11155111": 50400n,
  "1337": 50400n,
  "17000": 50400n,
  "412346": 50400n,
  "42161": 201600n,
  "421614": 50400n,
  "42170": 201600n,
  "8453": 1209600n,
  "84532": 302400n,
}
`;
